cranelift/zkasm_data/benchmarks/MVP.md

```diff
@@ -0,0 +1,141 @@
+
+#### Hot labels: executed labels
+
+To build a histogram of executed labels. This helps to identify:
+
+- Labels most frequently executed which can hint at optimizations of the original program that is compiled to zkASM.
+- By looking at the body of the labels in the generated zkASM, it can help to find instruction sequences for which peephole optimizations might have a high impact as they compound over many executions.
+
+## Aspirational insights to gain while instrumenting zkASM
+
+- Try to get an idea of how to instrument instructions such that they can be assigned to the label they are nested under.
+- Try to understand how call hierarchies are represented in cranelift.
+
+### Where to log?
+
+#### `Inst` execution
+
+In `Inst::emit()` [before matching](https://github.com/near/wasmtime/blob/9dc141e0181e5c1121cf1daa42d80ce26468757b/cranelift/codegen/src/isa/zkasm/inst/emit.rs#L406) the variant.
+
+#### Label execution
+
+- Execution of zkASM labels emitted for `Inst::Label` is handled via logs for `Inst` execution (see above).
+- Logs for labels corresponding to functions are inserted when the label is created in [generate_zkasm\(\)](https://github.com/near/wasmtime/blob/9dc141e0181e5c1121cf1daa42d80ce26468757b/cranelift/filetests/src/test_zkasm.rs#L220).
+
+### What to log?
+
+To build the histograms, execution counters according to the following schema shall be collected:
+
+```json
+{
+    "labels": {
+        "functions": {
+            "function_1": 1,
+            "function_2": 14
+        },
+        "others": {
+            "label_1_1": 4,
+            "label_1_2": 8,
+            "label_2_1": 4
+        }
+    },
+    "instructions": {
+        "AluRRR": 12,
+        "Load": 8,
+        "Store": 8
+    }
+}
+```
+
+### How to log
+
+#### Variant 1: `zkevm-proverjs` helpers can maintain state
+
+If `zkevm-proverjs` helpers can maintain state during the execution of a program, above data structure can live in state. Then instrumented zkASM would call helpers like:
+
+```js
+$${count("labels", "functions", "function_2")}
+
+$${count("instructions", null, "Load")}
+```
+
+Before program termination, the data structure is logged.
+
+#### Variant 2: `zkevm-proverjs` helpers cannot maintain state
+
+If state cannot be maintained, every execution of an instruction and entering of a label needs to be logged.
+
+```js
+$${log("executing Load")}
+$${log("entering label_1_2")}
+```
+
+Logs are then processed to produce above data structure.
+
+## Construct visualizations
+
+Histograms allow to get a quick overview of which instructions and labels are executed how often. The histograms will be generated as SVG files using the [`svg` crate](https://crates.io/crates/svg). Having visualizations as SVGs provides a number of advantages:
+
+- No dependencies required as SVG files can be viewed in a web browser.
+- They can be embedded in Github comments, markdown files and other documents.
```

Comment:
Another requirement here would be to make it easy to experiment with different visualizations. Looking through the API of SVG crate, it is quite low-level. So we would need to build tools on top to do even simple visualizations like histograms.

A simpler alternative might be to take a dedicated visualization library e.g. Seaborn/Matplotlib in Python and use the high-level API to build the visualizations we want. For example, snippet for histogram is very short: https://seaborn.pydata.org/examples/faceted_histogram.html 
To make this graph easy to view on GitHub, you can save it as "png" right from Seaborn: https://www.statology.org/save-seaborn-plot/

Reply:
Also, in case you are not very familiar with Seaborn/Matplotlib, llms are very effective in writing code using this libraries  (=